subprojects/normalization-java/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java

```diff
@@ -181,7 +181,11 @@ private HashCode fingerprintTreeFile(RegularFileSnapshot fileSnapshot) {
             if (shouldBeIgnored) {
                 return null;
             }
-            return classpathResourceHasher.hash(fileSnapshot);
+            if (ZipHasher.isZipFile(fileSnapshot.getName())) {
+                return zipHasher.hash(fileSnapshot);
+            } else {
+                return classpathResourceHasher.hash(fileSnapshot);
+            }
```

Comment:
Shall we do this as a separate PR? The rest of the changes are quite unambiguous, but this one smells like it could introduce some breaking changes, and would be nice to keep track of separately.

Reply:
Good point.  I've moved these changes to https://github.com/gradle/gradle/pull/13352.